File: Business_System_Simulation/Q2. Design Twitter/Q2_DesignTwitter.py
class User(object):
    def __init__(self,id):
        self.posts = {}                 # Stores user's tweets as {tweetId: time}
        self.following = {}             # Stores users this user follows {userId: User}
        self.id = id                    # Save the user's ID

    def addPost(self,tweetId,time):
        self.posts[tweetId] = time      # Add a tweet with its timestamp

    def addFollower(self,userFollowee):
        self.following[userFollowee.id] = userFollowee   # Add the followee to following list

    def deleteFollower(self, userFollowee):
        if userFollowee.id in self.following:            # If already following
            del self.following[userFollowee.id]          # Remove the followee


class Twitter(object):
    def __init__(self):
        self.users = {}                 # Stores all users {userId: User}
        self.ticks = 0                  # Global time counter for tweets

    def postTweet(self, userId, tweetId):
        if userId not in self.users:    # Create user if not exists
            user = User(userId)
            self.users[userId] = user

        self.users[userId].addPost(tweetId,self.ticks)   # Add tweet with current time
        self.ticks+=1                   # Increase global time


    def getNewsFeed(self, userId):
        if userId not in self.users:    # If user doesn't exist return empty feed
            return []

        user = self.users[userId]       # Get the user object
        own_follower_posts = user.posts.copy()  # Copy user’s own posts

        for fol in user.following.values():     # Add posts from all followees
            own_follower_posts.update(fol.posts)

        own_follower_posts = sorted(           # Sort by time (latest first)
            own_follower_posts.items(),
            key=lambda x:x[1],
            reverse=True
        )

        return [k for k,v in own_follower_posts[:10]]    # Return top 10 tweetIds


    def follow(self, followerId, followeeId):
        if followerId != followeeId:     # User cannot follow themselves
            if followerId not in self.users:     # Create follower if missing
                user_follower = User(followerId)
                self.users[followerId] = user_follower

            if followeeId not in self.users:     # Create followee if missing
                user_followee = User(followeeId)
                self.users[followeeId] = user_followee

            self.users[followerId].addFollower(self.users[followeeId])  # Add follow link


    def unfollow(self, followerId, followeeId):
        if followerId not in self.users:          # Create follower if missing
            user_follower = User(followerId)
            self.users[followerId] = user_follower

        if followeeId not in self.users:          # Create followee if missing
            user_followee = User(followeeId)
            self.users[followeeId] = user_followee

        if followeeId in self.users[followerId].following:  # If actually following
            self.users[followerId].deleteFollower(self.users[followeeId])  # Remove followee

        print(self.users[followerId].following)   # Print remaining following list


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: Business_System_Simulation/Q2. Design Twitter/Q2_DesignTwitter.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

class Twitter {

    private static int timeStamp = 0; // global time counter for ordering tweets
    private Map<Integer, Set<Integer>> followMap; // stores users each user follows
    private Map<Integer, List<int[]>> tweetMap;   // stores tweets per user as [time, tweetId]

    public Twitter() {
        followMap = new HashMap<>(); // initialize follow map
        tweetMap = new HashMap<>();  // initialize tweet map
    }

    // Post a new tweet by userId
    public void postTweet(int userId, int tweetId) {
        tweetMap.putIfAbsent(userId, new ArrayList<>()); // create list if not present
        tweetMap.get(userId).add(new int[]{timeStamp++, tweetId}); // store tweet with timestamp
    }

    // Get 10 most recent tweets in user's feed
    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]); // max-heap by time

        // Add user's own tweets
        if (tweetMap.containsKey(userId)) {
            for (int[] tweet : tweetMap.get(userId)) {
                pq.offer(tweet); // push to heap
            }
        }

        // Add tweets of followed users
        if (followMap.containsKey(userId)) {
            for (int followee : followMap.get(userId)) {
                if (tweetMap.containsKey(followee)) { // if followee has tweets
                    for (int[] tweet : tweetMap.get(followee)) {
                        pq.offer(tweet); // push to heap
                    }
                }
            }
        }

        // Collect up to 10 most recent tweets
        List<Integer> res = new ArrayList<>();
        int count = 0;
        while (!pq.isEmpty() && count < 10) {
            res.add(pq.poll()[1]); // get tweetId from heap
            count++;               // count tweets added
        }
        return res; // return final list
    }

    // Make follower follow followee
    public void follow(int followerId, int followeeId) {
        if (followerId == followeeId) return; // cannot follow yourself
        followMap.putIfAbsent(followerId, new HashSet<>()); // create set if not exists
        followMap.get(followerId).add(followeeId); // add followee to follower's set
    }

    // Make follower unfollow followee
    public void unfollow(int followerId, int followeeId) {
        if (followMap.containsKey(followerId)) { // if follower exists
            followMap.get(followerId).remove(followeeId); // remove followee
        }
    }
}

/**
 * Example usage:
 * Twitter obj = new Twitter();
 * obj.postTweet(1, 5);
 * List<Integer> feed = obj.getNewsFeed(1);
 * obj.follow(1, 2);
 * obj.unfollow(1, 2);
 */
